### PR TITLE
Buildkite: Cache UnrealEngine between runs

### PR DESCRIPTION
**Pull request by ca-johnson**
* Store UnrealEngine downloads in `UnrealEngine-Cache` directory
* gsutil cp with noclobber flag to allow skipping re-downloads (and resume aborted downloads)
* allow 7z to skip files that are already on disk (and resume aborted unzips)
* Junction link the correct unzipped version of UnrealEngine from the cache

This trims 5-7m from the baseline build time, which is a speedup of up to 3x

Unfortunately the full benefit wont be seen until *all* branches of spatialos/UnrealGDK include the `git clean` whitelisting of `UnrealEngine-Cache`

Also, we need to:
https://github.com/improbable/platform/pull/11919
- [x] Update baked-in buildkite agent hooks to not forcefully clean anyway 😓

### CODE REVIEW

**Comment by ca-johnson** [comment]
Unfortunately this work will be delayed because we bake a hook into agents that does a duplicate `git clean` in the pre-exit hook, but doesn't respect how the git clean plugin overrides these flags.

Need to get a change reviewed, rebake the GDK buildkite image and rollout to see any improvement.

**Comment by ca-johnson** [comment]
@m-samiec merging is blocked, not sure what I need to do to get this in

File: ci/get-engine.ps1 [hunk 1]
```diff
@@ -46,6 +50,10 @@ pushd "$($gdk_home)"
         }
     popd
 
+    ## Create an UnrealEngine symlink to the correct directory
+    Remove-Item "UnrealEngine" -ErrorAction ignore -Recurse -Force
```
Comment (m-samiec) [inline ci/get-engine.ps1]: Two things - why is this necessary? Should this folder be fresh from the checkout?
And shouldn't we prepend $gdk_home to the path or use the $unreal_path below directly?

ie. cmd /c mklink /J $unreal_path "UnrealEngine-Cache\$($unreal_version)"

**Comment by m-samiec** [comment]
@ca-johnson What's the status of `Update baked-in buildkite agent hooks to not forcefully clean anyway` ?
I assume it's not required for this PR as you're using your plugin (override?) of the hook instead?

File: ci/get-engine.ps1 (hunk 1 above)
Comment (ca-johnson) [reply ci/get-engine.ps1]: In practice, yes it will be deleted - this is mainly so you can iterate on the script without having to delete it by hand in between each run.

IMO "UnrealEngine" is more clear than "$gdk_home", since it's clearly a directory in the cwd, whereas $gdk_home could be anywhere (even though its actually just the cwd).

**Comment by ca-johnson** [comment]
Since the build is passing its safe to merge this PR

See https://github.com/improbable/platform/pull/11919, we basically have an extra clean step that forcefully cleans everything regardless of what the plugin says.

We just need to rebake the gdk-for-unreal image and retarget the build to these workers for this PR to have any impact.

At the minute, the build capability was failing because the dotnet3.5 package we were installing seems to have vanished from the internet. Do you know anything about this? I think we can remove it and the UnrealEngine + UnrealGDK builds still pass, but not sure about any of the spin-off repos with example games etc.

File: ci/get-engine.ps1 (hunk 1 above)
Comment (m-samiec) [reply ci/get-engine.ps1]: Thanks for clearing that up.

re: paths, I think because immediately after this we expect UnrealEngine to live in gdk_home, we should be making the symlink using it
